### PR TITLE
Refactor VideoGallery to have a component for 'default' layout

### PR DESCRIPTION
**Pull request by mgamis-msft**
# What

Created a `DefaultLayout` component that is used in VideoGallery that handles the layout of elements. The logic of what goes in grid, horizontal gallery will remain in VideoGallery because this logic will also be used in the floatingLocalVideo layout as well as the upcoming `PinnedLayout`. PR #2561 adds a `FloatingLocalVideoLayout` component so that code will VideoGallery will be reduced by ~100 lines.

# Why

https://skype.visualstudio.com/SPOOL/_workitems/edit/3054058

# How Tested

Passing enzyme tests. Tested on storybook and local Calling sample.

# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/react-components/src/components/ErrorBar.test.tsx [hunk 1]
```diff
@@ -3,19 +3,22 @@
```
Comment (mgamis-msft) [inline packages/react-components/src/components/ErrorBar.test.tsx:3]: Changes in this file are unrelated to VideoGallery. Just wanted to avoid the unregistered icon warnings in the console when running the tests.

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      data-ui-id={ids.videoGallery}
+      ref={containerRef}
+      className={mergeStyles(videoGalleryOuterDivStyle, styles?.root)}
+    >
+      <Stack horizontal={false} styles={videoGalleryContainerStyle}>
+        <GridLayout key="grid-layout" styles={styles?.gridLayout}>
+          {gridElements}
+        </GridLayout>
+        {horizontalGalleryElements && horizontalGalleryElements.length > 0 && (
+          <div style={{ paddingTop: '0.5rem' }}>
```
Comment (JamesBurnside) [inline packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:85]: try avoid raw `div` elements. Could we instead give the stack `tokens={{ childrenGap: '0.5rem' }}`

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  horizontalGalleryElements?: JSX.Element[];
+}
+
+/**
+ * DefaultLayout displays elements passed to `gridComponents` prop in the grid area and elements passed to
+ * `horizontalGalleryComponents` prop in the horizontal gallery area.
+ *
+ * @private
+ */
+export const _DefaultLayout = (props: _DefaultLayoutProps): JSX.Element => {
```
Comment (JamesBurnside) [inline packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:65]: nit: if this is `private` it doesn't nee the underscores:
```suggestion
export const DefaultLayout = (props: DefaultLayoutProps): JSX.Element => {
```

File: packages/react-components/src/components/VideoGallery.tsx [hunk 4]
```diff
@@ -426,72 +427,82 @@ export const VideoGallery = (props: VideoGalleryProps): JSX.Element => {
...
+            remoteScreenShareComponent
+          ) : localParticipant?.isScreenSharingOn ? (
+            localScreenShareStreamComponent
+          ) : (
+            <GridLayout key="grid-layout" styles={styles?.gridLayout}>
+              {gridTiles}
+            </GridLayout>
+          )}
+          {horizontalGalleryPresent && (
+            <div style={{ paddingTop: '0.5rem' }}>
```
Comment (JamesBurnside) [inline packages/react-components/src/components/VideoGallery.tsx:479]: nit:
As far as I can tell this is the exact same code as the defaultLayout.
I like the separation! And I like `ResponsiveHorizontalGallery` doesn't have these defaults and is kept my agnostic. But consider in `ResponsiveHorizontalGallery.tsx` adding a 
```tsx
/**
 * A ResponsiveHorizontalGallery styled for the @link{VideoGallery}
 */
const VideoGalleryResponsiveHorizontalGallery = (props): JSX.Element => (
   <ResponsiveHorizontalGallery
              key="responsive-horizontal-gallery"
              containerStyles={horizontalGalleryContainerStyle(false, isNarrow)}
              horizontalGalleryStyles={concatStyleSets(horizontalGalleryStyle(isNarrow), styles?.horizontalGallery)}
              childWidthRem={
                isNarrow ? SMALL_HORIZONTAL_GALLERY_TILE_SIZE_REM.width : LARGE_HORIZONTAL_GALLERY_TILE_SIZE_REM.width
              }
              buttonWidthRem={HORIZONTAL_GALLERY_BUTTON_WIDTH}
              gapWidthRem={HORIZONTAL_GALLERY_GAP}
            >
              {horizontalGalleryElements}
   </ResponsiveHorizontalGallery>
);
```

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    >
+      <Stack horizontal={false} styles={videoGalleryContainerStyle}>
+        <GridLayout key="grid-layout" styles={styles?.gridLayout}>
+          {gridElements}
+        </GridLayout>
+        {horizontalGalleryElements && horizontalGalleryElements.length > 0 && (
+          <div style={{ paddingTop: '0.5rem' }}>
+            <ResponsiveHorizontalGallery
+              key="responsive-horizontal-gallery"
+              containerStyles={horizontalGalleryContainerStyle(false, isNarrow)}
```
Comment (JamesBurnside) [inline packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:88]: containerStyles and horizontalGalleryStyles should be memoized

**Review by JamesBurnside — APPROVED** [review]
left some nits but looks good! Love these refactors and thanks for keeping the PRs well-isolated!

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+   * Elements to display in the horizontal gallery
+   */
+  horizontalGalleryElements?: JSX.Element[];
+}
+
+/**
+ * DefaultLayout displays elements passed to `gridComponents` prop in the grid area and elements passed to
+ * `horizontalGalleryComponents` prop in the horizontal gallery area.
+ *
+ * @private
```
Comment (dmceachernmsft) [inline packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:63]: Should this be internal? or are we going to export these layouts at some point?

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+   * @Example
+   * ```
+   * <_DefaultLayout styles={{ root: { border: 'solid 1px red' } }} />
+   * ```
+   */
+  styles?: DefaultLayoutStyles;
+  /**
+   * Elements to display in the grid
+   */
+  gridElements?: JSX.Element[];
```
Comment (dmceachernmsft) [inline packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:52]: Will the local video tile be included here? the default layout is the gridview right?

Comment (mgamis-msft) [reply packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:52]: DefaultLayout is a layout that taken in elements in the grid and in the horizontal gallery if any. Yes we are adding the local video tile in gridElements when the the layout of VideoGallery is 'default'

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx (hunk 6 above)
Comment (mgamis-msft) [reply packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:63]: We are not exporting `DefaultLayout` so this should just be private.

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    maxRemoteVideoStreams,
+    isScreenShareActive: !!screenShareComponent
+  });
+
+  let activeVideoStreams = 0;
+
+  const gridTiles = floatingLocalVideoLayout.gridParticipants.map((p) => {
+    return onRenderRemoteParticipant(
+      p,
+      maxRemoteVideoStreams
```
Comment (anjulgarg) [inline packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:82]: Would this logic fail if `maxRemoteVideoStreams` is set to 0?

File: packages/react-components/src/components/VideoGallery/videoGalleryUtils.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+          /* @conditional-compile-remove(PSTN-calls) */ /* @conditional-compile-remove(one-to-n-calling) */ !visibleCallingParticipantsSet.has(
+            p.userId
+          )
+      ) ?? [],
+    dominantSpeakers,
+    lastVisibleParticipants: visibleAudioParticipants.current,
+    maxDominantSpeakers: maxAudioDominantSpeakers
+  });
+
+  const getGridParticipants = (): VideoGalleryRemoteParticipant[] => {
```
Comment (anjulgarg) [inline packages/react-components/src/components/VideoGallery/videoGalleryUtils.ts:66]: Let's memoize this so that this hook doesn't cause unnecessary renders in the component it's used :)

File: packages/react-components/src/components/VideoGallery/videoGalleryUtils.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    return visibleVideoParticipants.current.length > 0
+      ? visibleVideoParticipants.current
+      : visibleAudioParticipants.current.concat(callingParticipants);
+    return visibleVideoParticipants.current.length > 0
+      ? visibleVideoParticipants.current
+      : visibleAudioParticipants.current;
+  };
+  const gridParticipants = getGridParticipants();
+
+  const getHorizontalGalleryRemoteParticipants = (): VideoGalleryRemoteParticipant[] => {
```
Comment (anjulgarg) [inline packages/react-components/src/components/VideoGallery/videoGalleryUtils.ts:77]: Let's memoize this so that this hook doesn't cause unnecessary renders in the component it's used :)

File: packages/react-components/src/components/VideoGallery/videoGalleryUtils.ts (hunk 9 above)
Comment (anjulgarg) [reply packages/react-components/src/components/VideoGallery/videoGalleryUtils.ts:66]: You can use React.useCallback

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+ */
+export const DefaultLayout = (props: DefaultLayoutProps): JSX.Element => {
+  const {
+    remoteParticipants = [],
+    dominantSpeakers,
+    localVideoComponent,
+    screenShareComponent,
+    onRenderRemoteParticipant,
+    styles,
+    maxRemoteVideoStreams,
```
[the remark this answers is not in the document]
Comment (mgamis-msft) [reply packages/react-components/src/components/VideoGallery/DefaultLayout.tsx:64]: We assign that default in VideoGallery prop which is passed to DefaultLayout. I don't want to have the same default value everywhere.

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx (hunk 8 above)
Comment (mgamis-msft) [reply packages/react-components/src/components/VideoGallery/DefaultLayout.tsx]: Yes. Fixed.

File: packages/react-components/src/components/VideoGallery/DefaultLayout.tsx (hunk 5 above)
Comment (mgamis-msft) [reply packages/react-components/src/components/VideoGallery/DefaultLayout.tsx]: Memoized. Thanks.

**Review by anjulgarg — APPROVED** [review]

**Review by dmceachernmsft — APPROVED** [review]
